Hands-On Lab: Cross-Validating a Model  
● Step 1: Take a Week 3 model and evaluate it with 5-fold cross-validation using cross_val_score.   
● Step 2: Report the mean and standard deviation of the scores across folds.    
● Step 3: Compare the cross-validated estimate to the single-split score from Day 1 and explain any 
difference.     
● Step 4: For a classification task, confirm stratified folds are used and explain why that matters here.   

Taking week3/day3 dataset (breast cancer). (week4/day1)

In [2]:
from sklearn.datasets import load_breast_cancer

from sklearn.model_selection import train_test_split

from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import accuracy_score , roc_auc_score, classification_report

from sklearn.model_selection import cross_val_score


load & process the dataset:

In [5]:
Br_can = load_breast_cancer()

X = Br_can.data
y = (Br_can.target == 0).astype(int)  # make it into: malignant = 1, benign = 0

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)


Cross validate:

In [19]:
RF = make_pipeline(StandardScaler(), RandomForestClassifier(n_estimators=100))

RF.fit(X_train,y_train)
RF_score = RF.score(X_test, y_test)

scores = cross_val_score(RF, X_train, y_train, cv = 5 , scoring= "f1") # 5 folds (cv = 5)

'''
scores = cross_val_score(RF, X, y, cv = 5 , scoring= "f1") # 5 folds (cv = 5)
we shouldn't use it directly with X, y becuase:
since it uses test data for model evaluation it creates data leakage, making it untrustworthy.
That is why X_train, y_train must be used.
'''

print(f"Single test split score: {RF_score:.3f}\n")
print(f"scores : {scores} \n")          # one score per fold
print(f"scores mean: {scores.mean():.3f} \n")   # the reliable average
print(f"scores std: {scores.std()} \n")    # how much the score varies across folds

Single test split score: 0.965

scores : [0.95238095 0.92307692 0.97142857 0.94117647 0.90625   ] 

scores mean: 0.939 

scores std: 0.022633654667152756 



Single test split score: 0.965

5 Folds cross validated scores mean: 0.939 

While there are difference in the scores, the cross validated is more trustworthy even if it is showing a samller value.  

This is due to its nature in taking multiple folds and taking each's score into a mean that then reflects the overall score better than just taking one part.   


because RF is a classifier and y_train is a classification target, 
 
 scikit-learn will normally use StratifiedKFold automatically when you pass an integer like cv=5.   

 cv = 5 acts as a StratifiedKFold.

Stratified K-fold keeps approximately the same class proportions in every fold as in the original training set.

 

A version with clear usage of stratified Folds :

In [ ]:
from sklearn.model_selection import StratifiedKFold

cv = StratifiedKFold(n_splits= 5 ,shuffle=True, random_state= 42 ) # cv folds the data into n_splits folds.

scores = cross_val_score(RF, X_train, y_train, cv = cv , scoring= "f1") # (cv = cv)


print(f"scores : {scores} \n")          # one score per fold
print(f"scores mean: {scores.mean():.3f} \n")   # the reliable average
print(f"scores std: {scores.std()} \n")    # how much the score varies across folds

scores : [0.98461538 0.93939394 0.88235294 0.92537313 0.9375    ] 

scores mean: 0.934 

scores std: 0.03269552359454083 

